doc/foreign-function-interface.md

```diff
@@ -0,0 +1,341 @@
+
+```haskell
+import GHCJS.Types (JSString, JSVal)
+
+foreign import javascript interruptible
+  "require('fs').stat($1, function (err, stat) { $c(err, stat); })" -- node.js API
+  js_fsStat :: JSString -> IO (JSVal, JSVal) -- IO (TypeOfErr, TypeOfStat)
+```
+
+### Size-2 types in Javascript FFI
+
+You won't need or want to use size-2 types in most cases, but it's explained here for completeness.
+In Javascript FFI, arguments of size-2 types are represented as `$1_1`, `$1_2`, `$2_1`, `$2_2`, ...
+Result of size-2 type is represented as `$r1` and `$r2`.
+
+For `Int64` and `Word64`, `$1_1` represents the high 32-bits of the first argument. `$1_2` represents the low 32 bits of the first argument. As you can see below, it's better to not use `Word64` and `Int64` in Javascript FFI unless you want to split them into two parts and manipulate the parts separately.
+
+```haskell
+foreign import javascript unsafe
+  "$r1=$1_1; $r2=$1_2" idInt64 :: Int64 -> Int64
+
+foreign import javascript unsafe
+  "$r1=$1_1; $r2=$1_2" idWord64 :: Word64 -> Word64
+```
+
+The pointer types(`Ptr a`, `StablePtr a`, and `FunPtr a`) are used mostly to interface with GHC constructs that depend on C code. For the pointer types, `$r1` is a javascript object, and `$r2` is an offset within the object. Examples of dealing with the pointer types are listed below.
+
+```haskell
+foreign import javascript unsafe "$r1 = $1; $r2 = 0;"
+  js_mkPtr :: JSVal -> Ptr a
+
+foreign import javascript unsafe "$r = $1_1;"
+  js_ptrVal  :: Ptr a -> JSVal
+```
+
+### Caveats on Javascript FFI
+
+* `Int64` and `Word64` can be manipulated naturally in haskell code, and `Int64` and
+`Word64` can be passed into and returned from javascript functions via
+Javascript FFI. However, since javascript doesn't support `Int64` and `Word64`,
+they are represented as size-2 numerical types in javascript, and it'll be very very difficult to manipulate size-2 numerical types in javascript functions. Therefore, it's *strongly* recommended that you use `Double` as 53-bit integer instead of `Int64` and `Word64` in Javascript FFI.
+* It's ok to pass `Float` values into Javascript FFI, and `Float` can be returned
+from Javascript FFI. But, since javascript has only `Double`. Javascript FFI
+can return `Double` values exceeding the limitation of `Float`. Thus, it's
+recommended that you return `Double` or `JSVal` instead of `Float` in Javascript FFI.
+* You won't need to use `Ptr a`, `FunPtr a`, or `StablePtr a` in Javascript FFI in most cases because they exist to interface with C code.
+
+# Calling Haskell From Javascript
+
+It is possible to call haskell from javascript. But, as far as I know, there is no clean way to call haskell from javascript, yet. It's better to not think about it for now. If you think calling GHCJS from javascript is doable, feel free to modify this section.
```

Comment:
This section should say something like.  "In order to call Haskell code from JavaScript the function you must first export the function."  Then it should discuss how to do this with https://github.com/ghcjs/ghcjs-base/blob/master/GHCJS/Foreign/Callback.hs